File: com/white/ReverseNodesInKGroup.java
package com.white;

import java.util.ArrayList;

public class ReverseNodesInKGroup {
	
	// 25. Reverse Nodes in k-Group
	
	public static ListNode reverseKGroup(ListNode head, int k) {
        
		if (head==null || k<=1) {
			return head;
		}
		
		ListNode dummy = new ListNode(0);
		ListNode groupHead = head;
		ListNode groupTail = dummy;
		ListNode curr = head;
		
		ArrayList<ListNode> kNodes = new ArrayList<>(k);
		
		while (true) {
			if (curr==null)
				break;
			
			for (int i = 0; i < k; i++) {
				kNodes.add(curr);
				if (curr.next == null) {
					curr = curr.next;
					break;
				}
				curr = curr.next;
			}

			if (kNodes.size() < k) {
				if (groupTail==dummy) {
					dummy.next = head;
				} else {
					groupTail.next = kNodes.get(0);
				}
				break;
			} else {
				groupHead = kNodes.get(k-1);
				ListNode temp = groupHead;
				for (int i=k-2; i>=0; i--) {
					temp.next = kNodes.get(i);
					temp = temp.next;
				}
				kNodes.clear();
				groupTail.next = groupHead;
				groupTail = temp;
				groupTail.next = null;
			}
		}
		
		return dummy.next;
    }
}

File: LeetcodeCore/ReverseNodesInKGroup.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace LeetcodeCore
{
    public class ReverseNodesInKGroup
    {
        // 25. Reverse Nodes in k-Group
        public ListNode ReverseKGroup(ListNode head, int k)
        {
            var dummy = new ListNode(0);
            dummy.next = head;

            var nodeA = dummy;
            var nodeB = head;
            while (true)
            {
                var currCount = k-1;
                while (currCount > 0 && nodeB != null)
                {
                    nodeB = nodeB.next;
                    currCount--;
                }

                if (currCount == 0 && nodeB != null)
                {
                    var nodeC = nodeB.next;
                    var results = ReverseGroup(nodeA.next, nodeB);
                    // connect this group with prev and latter nodes
                    nodeA.next = results.Item1;
                    results.Item2.next = nodeC;
                    // set for next group
                    nodeA = results.Item2;
                    nodeB = nodeC;
                }
                else
                {
                    return dummy.next;
                }
            }
        }

        private Tuple<ListNode, ListNode> ReverseGroup(ListNode head, ListNode tail)
        {
            var prev = (ListNode)null;
            var curr = head;
            while (!ReferenceEquals(curr, tail))
            {
                var temp = curr.next;
                curr.next = prev;
                prev = curr;
                curr = temp;
            }
            curr.next = prev;
            return new Tuple<ListNode, ListNode>(tail, head);
        }
    }
}
